rest/client/client.go

```diff
@@ -0,0 +1,211 @@
+	rc := resty.New()
+	rc.SetBaseURL(config.URL)
+	rc.SetAuthToken(config.Key)
+	rc.SetRetryCount(config.MaxRetries)
+	rc.SetTimeout(10 * time.Second)
+
+	return HTTPBase{
+		rc: rc,
+	}
+}
+
+func (b *HTTPBase) Call(method, url string, params Params, response json.Unmarshaler, opts ...Option) error {
+	req := b.newRequest(params, response, opts...)
+	res, err := req.Execute(method, url)
+	if err != nil {
+		return err
+	} else if res.IsError() {
+		return newErrorResponse(res)
+	}
+
+	return nil
+}
+
+func (b *HTTPBase) newRequest(params Params, response json.Unmarshaler, opts ...Option) *resty.Request {
+	options := mergeOptions(opts...)
+
+	req := b.rc.R().SetContext(options.Ctx)
+	if params != nil {
+		req.SetPathParams(params.Path())
+		req.SetQueryParams(params.Query())
+	}
+
+	if options.RequestID != nil {
+		req.SetHeader(HeaderRequestID, *options.RequestID)
+	}
+
+	if options.APIKey != nil {
+		req.SetAuthToken(*options.APIKey)
+	}
+
+	req.SetHeaderMultiValues(options.Headers)
+
+	req.SetResult(response).SetError(&BaseResponse{})
+
+	return req
+}
+
+// BaseResponse has all possible attributes that any response can use. It's intended to be embedded in a domain specific
+// response struct.
+// Ex:
+//		type User struct {
+//			Name string `json:"name"`
+//			Age  int    `json:"age"`
+//		}
+//
+//		type UserAPIResponse struct {
+//			apis.BaseResponse
+//			User User `json:"results,omitempty"`
+// 		}
+type BaseResponse struct {
+	Status    string `json:"status"`
+	RequestID string `json:"request_id"`
+	Count     int    `json:"count,omitempty"`
+	Error     string `json:"error,omitempty"`
+	Message   string `json:"message,omitempty"`
+
+	PaginationHooks
+}
+
+// PaginationHooks are links to next and/or previous pages.
+// Embed this struct into your API response if your endpoint is going to support pagination.
+type PaginationHooks struct {
+	NextURL     string `json:"next_url,omitempty"`
+	PreviousURL string `json:"previous_url,omitempty"`
+}
+
+// ErrorResponse is returned from the client if an undesirable status is returned.
+type ErrorResponse struct {
+	StatusCode int
+	Status     string
+	RequestID  string
+	Message    string
+}
+
+func newErrorResponse(res *resty.Response) *ErrorResponse {
+	statusErr := &ErrorResponse{
+		StatusCode: res.StatusCode(),
+		RequestID:  res.Header().Get(HeaderRequestID),
+	}
+
+	errRes := &BaseResponse{}
+	err := json.Unmarshal(res.Body(), errRes)
+	if err != nil {
+		return statusErr // always return status
+	}
+
+	statusErr.Status = errRes.Status
+	statusErr.Message = errRes.Error
+
+	if statusErr.RequestID == "" && errRes.RequestID != "" {
+		statusErr.RequestID = errRes.RequestID
+	}
+
+	return statusErr
+}
+
+func (e *ErrorResponse) Error() string {
+	return fmt.Sprintf("bad status with code: %d; message: %s; request ID: %s; internal status: %s", e.StatusCode, e.Message, e.RequestID, e.Status)
+}
+
+// Options are used to configure client calls.
+type Options struct {
```

Comment:
Calling this `Options` is interesting because it could create confusion with our Options Market Type